compiler/luci/pass/src/FuseGRUPass.cpp

```diff
@@ -0,0 +1,477 @@
+    if (less_node == nullptr)
+      return false;
+
+    luci::CircleNode *less_input;
+    if (not luci::fill(&less_input, &_less_const).with_commutative_args_of(less_node))
+      return false;
+
+    if (_less_const->dtype() != loco::DataType::S32)
+      return false;
+
+    if (_less_const->size<loco::DataType::S32>() != 1)
+      return false;
+
+    assert(_less_const->at<loco::DataType::S32>(0) > 0);
+  }
+
+  // 2 - Check while's input nodes
+  // Save hidden state input node
+  {
+    if (_while_node->input_count() != 5)
+      return false;
+
+    // Save input node
+    _ifm = dynamic_cast<luci::CircleNode *>(_while_node->input(4));
+    if (_ifm == nullptr)
+      return false;
+
+    _hidden_input = dynamic_cast<luci::CircleConst *>(_while_node->input(3));
+    if (_hidden_input == nullptr)
+      return false;
+  }
+
+  // 3 - check body graph
+  // TODO: add more checks
+  {
+    const auto body_graph = _while_node->body_graph();
+
+    if (loco::input_nodes(body_graph).size() != 5)
+      return false;
+
+    if (loco::output_nodes(body_graph).size() != 5)
+      return false;
+
+    const auto body_nodes = loco::active_nodes(loco::output_nodes(body_graph));
+
+    int32_t num_fc = 0;
+    int32_t num_split = 0;
+    int32_t num_logistic = 0;
+    int32_t num_mul = 0;
+    int32_t num_add = 0;
+    int32_t num_sub = 0;
+    int32_t num_reshape = 0;
+    int32_t num_gather = 0;
+    int32_t num_tanh = 0;
+    int32_t num_split_out = 0;
+
+    for (auto node : body_nodes)
+    {
+      auto circle_node = dynamic_cast<luci::CircleNode *>(node);
+      switch (circle_node->opcode())
+      {
+        case luci::CircleOpcode::CIRCLECONST:
+        case luci::CircleOpcode::CIRCLEINPUT:
+        case luci::CircleOpcode::CIRCLEOUTPUT:
+        case luci::CircleOpcode::CIRCLEOUTPUTEXCLUDE:
+          break;
+        case luci::CircleOpcode::FULLY_CONNECTED:
+          num_fc++;
+          break;
+        case luci::CircleOpcode::SPLIT:
+          num_split++;
+          break;
+        case luci::CircleOpcode::LOGISTIC:
+          num_logistic++;
+          break;
+        case luci::CircleOpcode::MUL:
+          num_mul++;
+          break;
+        case luci::CircleOpcode::ADD:
+          num_add++;
+          break;
+        case luci::CircleOpcode::SUB:
+          num_sub++;
+          break;
+        case luci::CircleOpcode::RESHAPE:
+          num_reshape++;
+          break;
+        case luci::CircleOpcode::GATHER:
+          num_gather++;
+          break;
+        case luci::CircleOpcode::TANH:
+          num_tanh++;
+          break;
+        case luci::CircleOpcode::CIRCLESPLITOUT:
+          num_split_out++;
+          break;
+        default:
+          return false;
+      }
+    }
+
+    if (num_fc != 2 or num_mul != 3 or num_logistic != 2 or num_split != 2 or num_add != 6 or
+        num_gather != 1 or num_reshape != 1 or num_sub != 1 or num_tanh != 1 or num_split_out != 6)
+      return false;
```

Comment:
I don't think counting elements can provide correct condition.
And tracking the graph for exact Op type looks laborious job.
But we may have to do with some good graph traversal codes.

Reply:
Okay, I will change it)